Leos gripper

Leo-Dan opened the pull request:
I made a launch file for the gripper
I implemented a script to control the gripper with the Xbox controller.
I will modify the Xbox controller mapping diagram according to the changes im proposing in there.
Will merge when UI is ready.

Leo-Dan committed:
integration of the gripper for ovis robotic arm

* I made a launch file to launch the gripper node easily
* Add a script for controlling the Xbox controller
* gitignore update
* remove the .vscode file form the repo

Leo-Dan committed:
clean up of the launch file

Leo-Dan committed:
clean up CMakeLists

Leo-Dan committed:
clean up joy_controller

Leo-Dan committed:
Button Y is for activating the gripper at startup of the node

lvanasse commented on robotiq_2f_gripper_control/launch/robotiq_gripper.launch:
```diff
@@ -0,0 +1,14 @@
+
+  <arg name="joy_dev" default="/dev/input/js0" />
+
+  <node pkg="joy" type="joy_node" name="joy_node">
```
I'm not sure that we want to launch the joy node at the same time as the gripper node. The computer in the robot will not have a controller connected to it when running the gripper node.

Leo-Dan replied:
I will remove the joy node from the launch file.
I will rename the Robotiq2fGripperRtuNode to a more generic name like Robotiq2FGripperNode and implement the controls in it directly.
Once the UI part is done, the Robotiq2FGripperNode will only subscribe to the OvisGripperAction topic.
I will make a msg file for controlling the gripper just like it is for Ovis.

Leo-Dan replied:
In fact I dont have to wait the UI is done, I can do it now

lvanasse replied:
:+1:  Sound perfect to me! Once you're done, you can resolve the comment :).

lvanasse commented on robotiq_2f_gripper_control/launch/robotiq_gripper.launch:
```diff
@@ -0,0 +1,14 @@
+    <param name="autorepeat_rate" value="50" />
+  </node>
+
+  <node pkg="robotiq_2f_gripper_control" type="Robotiq2FGripperRtuNode.py" name="Robotiq2FGripperRtuNode" output="screen" args="/dev/ttyUSB0" />
```
Do we need to have the two nodes `Robotiq2FGripperRtuNode` and `joy_controller` in the launch ?

Leo-Dan replied:
No we dont and we wont in a near future.

lvanasse commented on robotiq_2f_gripper_control/nodes/joy_controller.py:
```diff
@@ -0,0 +1,52 @@
+
+def joy_listener():
+    global command
+    rospy.init_node("joy_controller", anonymous=True)
```
Is `joy_controller` a good name for the node controlling the gripper with the Xbox controller? What do you think? Something related to the gripper should be in the name, maybe `robotiq_gripper` simply. Maybe add that we are using the modbus rtu ?

Leo-Dan replied:
the joy controller node will disappear with future integration actions.

lvanasse commented on robotiq_2f_gripper_control/nodes/joy_controller.py:
```diff
@@ -0,0 +1,52 @@
+def joy_listener():
+    global command
+    rospy.init_node("joy_controller", anonymous=True)
+    rospy.Subscriber("/joy", Joy, joy_callback, queue_size=1)
```
Also, I'm really not a fan just listening to the joy topic. Because what if there's another node listening to the joy topic ? Does your node have a way to distinguish between a topic meant for another node? 

I think we could add a ROS msg with only what we need to control the gripper and send it from the UI. This will make sure that only this node receives the topic.

Leo-Dan replied:
I agree with you it will be done that way

lvanasse commented on robotiq_2f_gripper_control/nodes/joy_controller.py:
```diff
@@ -0,0 +1,52 @@
+
+    if button_Y:
+        command = outputMsg.Robotiq2FGripper_robot_output()
+        command.rACT = 1  #activates the gripper 
```
As discussed previously, I don't think using magics numbers and a comment beside it is an efficient way to convey the information.

What if we reuse the `command.rACT = 1` somewhere else in the code? Do we add a comment beside it also? What if the value is used as a function argument? 

Simply changing it to a variable with a meaningful name would solve the issue.

Leo-Dan replied:
variables with significative names will be added.

saxtot replied:
Can we actually change these names? We send the robotiq-made msg to the gripper class, meaning it expects these specific message names, no?

lvanasse commented on robotiq_2f_gripper_control/nodes/joy_controller.py:
```diff
@@ -0,0 +1,52 @@
+    
+    elif button_X:
+        command.rPR = 0 #open
+   
```
Are the newline necessary here? If they are not, you can leave one newline to help ventilate the code, but here I think there's too many.  I think this comment apply for the rest of the code also.

Leo-Dan replied:
I will reformat the code so that it is nice to read and ventilated.

lvanasse replied:
There should be formatter/linter for Python in VSCode that might help you with that.

benmalenfant commented on robotiq_2f_gripper_control/launch/robotiq_gripper.launch:
```diff
@@ -0,0 +1,14 @@
+
+  <arg name="joy_dev" default="/dev/input/js0" />
+
+  <node pkg="joy" type="joy_node" name="joy_node">
```
On devrait pas lancer la node joy directement a partir du gripper, sa fonctionne bien en standalone mais dans le robot, c'est le UI qui s'occupe de publier joy, on pourrait garder ce launch file comme un standalone launchfile pour developpement mais sa prenderais un autre launchfile ou on ne lance pas joy_node pour le déploiement dans le robot

Leo-Dan replied:
Effectivement tout cela était pour faciliter les test. Pour l'intégration avec le reste du système un nouveau launch file sera créer.  voir les commentaires que j'ai laissé à Ludo plus haut pour voir tout des changements qui seront fait.

lvanasse reviewed (changes requested):
Tell me if I'm wrong, but from my understanding, this node allows for controlling the gripper with a joy topic. If that's true, and it is the only reason, can we implement this feature in the original node (`Robotiq2FGripperNode`) instead? 

Also, I'm not convinced that the node should listen to the whole joy topic. We should use a specific topic, like we did in the ovis' arm node. This will allow for better segmentation of what is sent from the UI to the node.

benmalenfant reviewed (commented):
Vraiment nice! Faudrait juste enlever le joy dans le launchfile pour éviter les conflits avec le UI

Also, pour le futur, ce serrait p-e nice d'avoir une node commune aux projets qui utilisent le joystick qui s'occupe de dispatch les infos ou sinon on peut faire comme avant pis avoir tt la logique dans le ui

saxtot committed:
:zap: Change launch file logic to allow launch without joy

saxtot committed:
:zap: Add launch file to launch gripper with joy on its own

saxtot committed:
:adhesive_bandage: Correct simplest launch file

saxtot committed:
:hammer: Add initial version of single node control of gripper for testing

Do not merge, contains an unnecessary additional publisher for test logging
and other artefacts of the two intial nodes

saxtot committed:
:memo: Update CMake for renamed single node

saxtot committed:
:coffin: Adjust .vscode value in .gitignore

saxtot committed:
:construction: Clean up and continued work on gripper node

saxtot committed:
:art: Change repo structure to have a specific ovis_gripper package

saxtot committed:
:art: Add necessary package structure with renamed files

saxtot committed:
:package: Add required build dependencies to package

saxtot committed:
:pencil2: Update file names to be closer to normal .msg standards

saxtot committed:
:label: Add robot namespace to launch file

saxtot committed:
:hammer: Update CMakeLists to match new .msg names

saxtot committed:
:label: Change topic names to match convention and update msg names

saxtot committed:
:speech_balloon: Add custom msg type

saxtot committed:
:construction: Integrate new msg type

saxtot committed:
:rewind: Restore robotiq_2f_gripper_control package to initial value

saxtot committed:
:package: Add new msg to CMakeLists

saxtot committed:
:hammer: Fix msg typo and build

saxtot committed:
:art: Format control code

saxtot committed:
:memo: Add msg fields use to msg files

saxtot committed:
:wrench: Adujst control loop to give more flexibility and refresh the command

saxtot committed:
:coffin: Remove launch file including the joy node

saxtot committed:
:white_check_mark: Add comments and change redundant variable name

saxtot committed:
:sparkles: Initialization correction and conclusive tests

Message may change depending on UI tests

saxtot committed:
:sparkles: Add boot sequence and change control logic for controller use

saxtot committed:
:sparkles: Add send_command function to allow boot sequence

saxtot committed:
:wrench: Adjust message type read and its values

saxtot committed:
:zap: Fix boot problems

saxtot committed:
:heavy_plus_sign: Add new msg to CMakeLists

saxtot committed:
:art: Improve code comments and add constants usage

saxtot committed:
:ambulance: Switch from constants in a msg to a Python file

saxtot committed:
:zap: Add constant file

saxtot commented:
So! I've went and used this project to learn a lot about rospy, so this isn't 100% clean and multiple commits have been overwritten.
To facilitate the PR, I'll try and list the most notables points.
- Usage: It currently works with the UI/controller with the B button when in ARM mode. Press once to close, press again to open
- Message: I use a simple custom msg. The callback loop uses 0 and 1 for debugging and 2 for actual movement.
- Topics: The node publishes two topics. A publisher for the requests sent to the gripper and the other for the information the gripper sends back. The node subscribes to the third topic, the one the UI publishes on with the custom msg.
- Package: For simplicity and harmony's sake, I have moved the node to a new package. It makes it a whole lot easier to know which launch file to use for us and allows us to easily pull anything Robotiq might change on their original fork.
- Boot: The gripper requires an initial sequence to be sent for it to work. I have tried multiple times to add it before the while() loop in some kind of boot sequence, but it seems to cut all subsequent communications with the gripper itself. It might be worth checking out in the future. For now,  I just make it so every message sent as the setup arguments added by default and it works well. The only side effect is that this replaces the boot loop of the gripper (closing and opening once) and replaces it with a quarter of a second stutter before its first action (only the first one to be clear).
- Constants: As Python doesn't easily allow constants cleanly, I have added a second file for them.
- Previous PR comments: I believe I have addressed all previous comments as of this message. Be sure to verify that it is true.

As such, I'm requesting a new review from all of you!

lvanasse commented on ovis_robotiq_gripper/CMakeLists.txt:
```diff
@@ -0,0 +1,68 @@
+target_link_libraries(robotiq_2f_gripper_ethercat_node_for_ovis
+   ${robotiq_ethercat_LIBRARIES}
+   ${catkin_LIBRARIES}
+  
```
Nitpick: Unnecessary newline

lvanasse commented on ovis_robotiq_gripper/launch/robotiq_gripper.launch:
```diff
@@ -0,0 +1,5 @@
+<launch>
+  <arg name="robot_namespace" default="/ovis/gripper" />
+  <node name="ovis_robotiq_gripper" pkg="ovis_robotiq_gripper" type="ovis_gripper_control.py" ns="$(arg robot_namespace)" output="screen" args="/dev/ttyUSB0"></node>
```
1. You don't have to add a closing tag if there's nothing in between.
2. I think we ought to have the `args="/dev/ttyUSB0"` as an args outside the node tag.

saxtot replied:
Agreed, corrected as of now.

lvanasse commented on ovis_robotiq_gripper/launch/robotiq_gripper.launch:
```diff
@@ -0,0 +1,5 @@
+<launch>
+  <arg name="robot_namespace" default="/ovis/gripper" />
+  <node name="ovis_robotiq_gripper" pkg="ovis_robotiq_gripper" type="ovis_gripper_control.py" ns="$(arg robot_namespace)" output="screen" args="/dev/ttyUSB0"></node>
+
```
Nitpick: Unnecessary newline

lvanasse commented on ovis_robotiq_gripper/msg/OvisGripper_robot_input.msg:
```diff
@@ -0,0 +1,8 @@
+uint8 gACT #getActivation
```
Let me know if that's possible or not, could we simply use the commented section at the variable instead of an obscure name? Using `getActivation` instead of `gACT`?

lvanasse commented on ovis_robotiq_gripper/msg/OvisGripper_robot_output.msg:
```diff
@@ -0,0 +1,6 @@
+uint8 rACT #requestActivation
```
Same thing as my previous comment here

McCaroon replied:
Je suis d'accord avec ca!
Il y en a qui sont pas mal "to the point" exemple: "gPO = getPosition" mais il en ont qui sont plus dur a se rapeller même après avoir lu la datasheet et les commentaires que t'a laissé.

saxtot replied:
It would be technically possible. It would also be a good deal of work, since these messages are the input type in their gripper object and the other used packages.
I do not think this change would currently be worth the time investment.

lvanasse replied:
@saxtot and I spoke face to face. And I must agree with keeping it this way. It will not be a node that's going to have a lot of work done. And if changing it doesn't really add much value, except for maintainability, I think we can keep it like this and move on since we have more important projects to do.

lvanasse commented on ovis_robotiq_gripper/nodes/ovis_gripper_control.py:
```diff
@@ -0,0 +1,131 @@
+    elif input.position == 2:
+        command.rPR = cons.OPEN  # open
+
+def mainLoop(device):
```
Should be `main_loop` if we are following the http://wiki.ros.org/PyStyleGuide

lvanasse commented on ovis_robotiq_gripper/nodes/ovis_gripper_control.py:
```diff
@@ -0,0 +1,131 @@
+from ovis_robotiq_gripper.msg import _OvisGripperPosition as positionMsg
+from ovis_robotiq_gripper.msg import _OvisGripper_robot_output as outputMsg
+from ovis_robotiq_gripper.msg import _OvisGripper_robot_input as inputMsg
+import gripper_constants as cons
```
Nitpick: You don't have to do it, but I when right by this line, and then I was wondering "What the heck is `cons`", maybe `const` or `constants` might be more clear, what do you think?

saxtot replied:
I will remove completely the "as" clause. As such, gripper_constants is as clear is it gets and not that long.

lvanasse commented on ovis_robotiq_gripper/nodes/ovis_gripper_control.py:
```diff
@@ -0,0 +1,131 @@
+
+    rospy.init_node("ovis_robotiq_gripper")
+
+    # The Gripper status is published on the topic named 'gripper_input'
```
If that's the status of the gripper, could we add something that this topic is about the status? Something like : "gripper_input_status"?

lvanasse replied:
Also `pub_input` instead of `pubInput`, I won't repeat it every time in the review, but you might want to do a full checkup

lvanasse commented on ovis_robotiq_gripper/nodes/ovis_gripper_control.py:
```diff
@@ -0,0 +1,131 @@
+        'gripper_input', inputMsg.OvisGripper_robot_input, queue_size=1)
+
+    # The Gripper command is published on the topic named 'gripper_output'
+    pubOutput = rospy.Publisher(
```
Same thing here, I'm not sure why we are using `out` instead of `command` or `cmd`?

lvanasse commented on ovis_robotiq_gripper/nodes/ovis_gripper_control.py:
```diff
@@ -0,0 +1,131 @@
+    pubOutput = rospy.Publisher(
+        'gripper_output', outputMsg.OvisGripper_robot_output, queue_size=1)
+
+    #boot_sequence(gripper, pubOutput)
```
Is that old code or a comment?

saxtot replied:
Old code, my bad

lvanasse commented on ovis_robotiq_gripper/nodes/ovis_gripper_control.py:
```diff
@@ -0,0 +1,131 @@
+        current_position = status
+
+        # Get the most recent command
+        rospy.Subscriber("/ovis/gripper/position_goal",
```
I'm not necessary familiar with ROS Python, is that the idiomatic way of doing it in Python? If I read it correctly, every time we loop we subscribe to a topic, couldn't this be done before, outside the while loop?

saxtot replied:
If I understand correctly, the standard way for a publisher is the while() loop and for a subscriber its rospy.spin().
In a combined node it becomes blurry. The end point of both are functionally the same. 
To use rospy.spin(), I would have to put the publishers in the subscriber callbacks and they would be conditional to said callback, making debugging less convenient. 

Sources: http://wiki.ros.org/rospy/Overview/Publishers%20and%20Subscribers and https://get-help.robotigniteacademy.com/t/what-is-difference-between-rospy-spin-and-while-not-rospy-is-shutdown/1849/5

lvanasse replied:
> The end point of both are functionally the same.

Humm I'm not convince, publishing see publish data to a topic to a certain rate. Subscriber wait to receive a published topic. The Subscriber doesn't require a rate like the publisher. So having it inside the loop doesn't make sense to me. 

Can you try and see what happen if you put it outside the loop? This will test my theory, if you have communication issues with the subscriber we keep it inside the `main_loop` and if it doesn't affect it, we'll keep it outside the loop.

Also, much like the problem we had with the E-Stop node where it would take a lot of CPU, we ought to add a rate also for this node.

lvanasse commented on ovis_robotiq_gripper/src/ovis_robotiq_gripper/__init__.py:
```diff
@@ -0,0 +1 @@
+#autogenerated by ROS python message generators
```
I'm not sure if this should be inside the git repo. What do you think? Is there's more info online? Because, if that's auto-generated, I'm assuming at build time, we don't need to keep it in the VCS.

saxtot replied:
It is required according to https://docs.python.org/3/tutorial/modules.html#packages

lvanasse reviewed (changes requested):
Ah, I see, this is a whole new subpackage inside the robotiq package. That's OK since this is already done, but what I tried to explain last time, was that we could simply add a node inside the `robotiq_2f_gripper_control/nodes/` folder. Because although, this is more clear that it is our subpackage, there's code duplication, and we can't really piggyback on the existing documentation. 

Anyway, good job, although some small changes would be good.

lvanasse commented:
> * Package: For simplicity and harmony's sake, I have moved the node to a new package. It makes it a whole lot easier to know which launch file to use for us and allows us to easily pull anything Robotiq might change on their original fork.

Ah, I haven't read that part before reviewing. Yeah, I more or less convinced on that one. But since it's already done, it doesn't really matter.

lvanasse commented:
> * Topics: The node publishes two topics. A publisher for the requests sent to the gripper and the other for the information the gripper sends back. The node subscribes to the third topic, the one the UI publishes on with the custom msg.

Why do we have to republish the request sent to the gripper if the UI already publish it ?

saxtot commented:
> > * Topics: The node publishes two topics. A publisher for the requests sent to the gripper and the other for the information the gripper sends back. The node subscribes to the third topic, the one the UI publishes on with the custom msg.
> 
> Why do we have to republish the request sent to the gripper if the UI already publish it ?

We had decided in the past that the UI should only publish inputs and not do any work. With this in mind, we made it so the work needed to build the necessary message for the gripper and alternate open-close were made by the node.
Also, note that no one currently subscribes to either the input/status or the output/command topics. 
The message built by the node is sent directly to the gripper object.

McCaroon reviewed (commented):
Ready to merge when change requests done.
Other than that, looks good! 
Works and tested it physically.

saxtot committed:
:construction: Improve package quality

saxtot committed:
:art: Improve naming patterns

saxtot commented:
For reference, I have discussed some of Ludo's comments separately with him. Since, originally, we only added to one of Robotiq's packages, I only built upon their code. However, having now a separate package, their disregard for naming conventions looks like mine. So I will resolve most of his comments with only a thumbs up or explain some things more in detail.

lvanasse reviewed (commented):
Everything looks good. Before I approve, I would just like you to do a small test for me about the subscriber inside the main loop. 

Once this is tested, I'll approve it :).

lvanasse commented:
> > > * Topics: The node publishes two topics. A publisher for the requests sent to the gripper and the other for the information the gripper sends back. The node subscribes to the third topic, the one the UI publishes on with the custom msg.
> > 
> > 
> > Why do we have to republish the request sent to the gripper if the UI already publish it ?
> 
> We had decided in the past that the UI should only publish inputs and not do any work. With this in mind, we made it so the work needed to build the necessary message for the gripper and alternate open-close were made by the node. Also, note that no one currently subscribes to either the input/status or the output/command topics. The message built by the node is sent directly to the gripper object.

I'm sorry, but I fail to understand the explanation here. Why do we publish the topic that we just received from the node? Especially since nobody is using it, why even have this publisher?

saxtot committed:
:sparkles: Change msg fields for output and remove subscriber from loop

lvanasse commented:
Awesome! Thank you for taking the time to do the test :)!
